qemu/softmmu/virtual_tlb.c

```diff
@@ -0,0 +1,106 @@
+#include <stdint.h>
+#include "qemu/osdep.h"
+#include "qemu-common.h"
+#include "exec/exec-all.h"
+#include "uc_priv.h"
+
+#include <stdio.h>
+
+static void raise_mmu_exception(CPUState *cs, target_ulong address,
+                                int rw, uintptr_t retaddr)
+{
+    cs->uc->invalid_error = 1;
+    cpu_exit(cs->uc->cpu);
+    cpu_loop_exit_restore(cs, retaddr);
+}
+
+static uc_mem_type rw_to_mem_type(int rw)
+{
+    switch (rw) {
+    case MMU_DATA_LOAD:
+        return UC_MEM_READ;
+    case MMU_DATA_STORE:
+        return UC_MEM_WRITE;
+    case MMU_INST_FETCH:
+        return UC_MEM_FETCH;
+    default:
+        return UC_MEM_READ;
+    }
+}
+
+static int perms_to_prot(int perms)
+{
+    int ret = 0;
+    if (perms & UC_PROT_READ) {
+        ret |= PAGE_READ;
+    }
+    if (perms & UC_PROT_WRITE) {
+        ret |= PAGE_WRITE;
+    }
+    if (perms & UC_PROT_EXEC) {
+        ret |= PAGE_EXEC;
+    }
+    return ret;
+}
+
+bool unicorn_fill_tlb(CPUState *cs, vaddr address, int size,
```

Comment:
If we go through this subroutine, we are bypassing all architectures' mmu implementation, correct?

Reply:
Yes, or more correct we call this subroutine instand of the architectures' mmu implementation.